src/codegen/yul/statements.rs

```diff
@@ -0,0 +1,443 @@
+use crate::ast::{Namespace, RetrieveType, Type};
+use crate::codegen::cfg::{ControlFlowGraph, Instr};
+use crate::codegen::statements::LoopScopes;
+use crate::codegen::vartable::Vartable;
+use crate::codegen::yul::builtin::process_builtin;
+use crate::codegen::yul::expression::{expression, process_function_call};
+use crate::codegen::{Expression, Options};
+use crate::sema::yul::ast;
+use crate::sema::yul::ast::{YulStatement, YulSuffix};
+use num_bigint::BigInt;
+use num_traits::FromPrimitive;
+use solang_parser::pt;
+use solang_parser::pt::StorageLocation;
+
+/// Transform YUL statements into CFG instructions
+pub(crate) fn statement(
+    yul_statement: &YulStatement,
+    contract_no: usize,
+    loops: &mut LoopScopes,
+    ns: &Namespace,
+    cfg: &mut ControlFlowGraph,
+    vartab: &mut Vartable,
+    early_return: &Option<Instr>,
+    opt: &Options,
+) {
+    if !yul_statement.is_reachable() {
+        return;
+    }
+
+    match yul_statement {
+        YulStatement::FunctionCall(_, _, func_no, args) => {
+            let returns = process_function_call(*func_no, args, contract_no, vartab, cfg, ns, opt);
+            assert_eq!(returns.len(), 1);
+            assert_eq!(returns[0], Expression::Poison);
+        }
+
+        YulStatement::BuiltInCall(loc, _, builtin_ty, args) => {
+            let expr = process_builtin(loc, builtin_ty, args, contract_no, ns, vartab, cfg, opt);
+            assert_eq!(expr, Expression::Poison);
+        }
+
+        YulStatement::Block(block) => {
+            for item in &block.body {
+                statement(item, contract_no, loops, ns, cfg, vartab, early_return, opt);
+            }
+        }
+
+        YulStatement::VariableDeclaration(loc, _, vars, init) => {
+            process_variable_declaration(loc, vars, init, contract_no, ns, cfg, vartab, opt);
+        }
+
+        YulStatement::Assignment(loc, _, lhs, rhs) => {
+            process_assignment(loc, lhs, rhs, contract_no, ns, cfg, vartab, opt)
+        }
+
+        YulStatement::IfBlock(_, _, condition, block) => process_if_block(
+            condition,
+            block,
+            contract_no,
+            loops,
+            ns,
+            cfg,
+            vartab,
+            early_return,
+            opt,
+        ),
+
+        YulStatement::Switch { .. } => {
+            // Switch statements should use LLVM switch instruction, which requires changes in emit.
+            unimplemented!("Switch statements for yul are not implemented yet");
```

Comment:
please don't make solang crash

Reply:
Solang will never crash here, because there is an error in sema.